hnn_core/network_builder.py

```diff
@@ -298,15 +295,20 @@ def _build(self):
 
         self._clear_last_network_objects()
 
-        self._gid_assign()
+        # initialise vectors for synaptic currents based on simulation params
+        n_times = np.arange(0., self.net.params['tstop'],
+                            self.net.params['dt']).size + 1
 
         # Create a h.Vector() with size 1xself.N_t, zero'd
         self.current = {
-            'L5Pyr_soma': h.Vector(self.net.n_times, 0),
-            'L2Pyr_soma': h.Vector(self.net.n_times, 0),
+            'L5Pyr_soma': h.Vector(n_times, 0),
+            'L2Pyr_soma': h.Vector(n_times, 0),
         }
 
+        self._gid_assign()
```

Comment:
I think `_gid_assign` should happen after `_create_cells_and_feeds`, no?

Reply:
To my surprise too: No! Turns out L398 is the key. The for-loop there must be on `self.net._gid_list`, which contains precisely those cells (gids) that exist on _this rank/host_. So the `_gid_assign` (note the round-robin for-loop) essentially 'picks out' the gids that belong to the present rank, after which `create_...` instantiates them _here_.

Reply:
But if cells are created _after_ the GIDs are assigned to different nodes, how will we allow users to add feeds? User code should not depend on how the GID assignment is done ...

Reply:
They can't. But that should be OK, no? We can create pretty visualisations of the input to a `net` before it ever gets built. So seeing the effects of adding and removing feeds (histograms etc.) can be done interactively, `_gid_assign` is not run until `Builder`. Note that the reliance on `self.net._p_unique` is something that the future implementation could make prettier, but right now it works as it is.

Reply:
Just to clarify, @cjayb it sounds like what you're saying is that all aspects of the feed/network that we intend the user to interact with should be defined prior to `NetworkBuilder`? If so, this makes sense since the nrn interface should be invisible to the user.

Reply:
Yes. That wasn't really my goal but it mostly just became so when extracting the feed instantiation into `Network`. Now I think it's a good goal though, as you say ;)

Reply:
Note that I'm not implying that GID handling is "solved" quite yet. Baby steps... One of them being the protected cell attribute.